File: HGCalMaskResolutionAna/python/Calibration.py
from UserCode.HGCalMaskResolutionAna.PartialWafersStudies import PartialWafersStudies

class Calibration(PartialWafersStudies, object):
    def __init__(self, flags):
        import numpy as np
        from collections import OrderedDict

        super(Calibration, self).__init__(flags)
        self.calib = OrderedDict({'L0': {}, 'L1': {}, 'L2': {}})
        self.mingenen = self.flags.mingenen
        self.etas_l = self.etaregions[:-1]
        self.etas_r = np.roll(self.etaregions, shift=-1)[:-1]
        self.label = self.flags.plotLabel
        self.samples = self.flags.samples
        self.mask = str(self.flags.mask)
        self.outpath = self.flags.outpath

    def _calibrate_spectrum(self, h, title, proc, func='pol1', plot=True):
        """
        Calibrates a DeltaE/E versus x spectrum based on the profile.
        """
        import os
        from UserCode.HGCalMaskResolutionAna.RootTools import buildMedianProfile

        prof = buildMedianProfile(h)
        prof.Fit(func)
        calibGr = prof.GetListOfFunctions().At(0).Clone(h.GetName()+title+'_calib')

        if plot:
            from ROOT import TCanvas, TLatex
            c = TCanvas('c','c',500,500)
            c.SetTopMargin(0.05)
            c.SetBottomMargin(0.1)
            c.SetLeftMargin(0.12)
            c.SetRightMargin(0.1)
            h.Draw('colz')
            prof.Draw('e2p')

            tex = TLatex()
            tex.SetTextFont(42)
            tex.SetTextSize(0.04)
            tex.SetNDC()
            tex.DrawLatex(0.12,0.96,'#bf{CMS} #it{simulation preliminary}')
            tex.DrawLatex(0.15,0.88,title)
            tex.SetTextAlign(31)
            tex.DrawLatex(0.97,0.96,proc)        
            c.SaveAs(os.path.join(self.outpath,h.GetName()+'_'+title+'.png'))
            prof.Delete()        
            c.Delete()

        return calibGr

    def _energies_for_calibration(self, data, ireg):
        """
        Returns an array of reconstructed energies, eta and average noise
        """
        import numpy as np
        x = [[] for _ in range(len(self.etas_l))]
        for i in range(0, data.GetEntriesFast()):
            data.GetEntry(i)
            genen = getattr(data,'genen')
            if genen < self.mingenen: 
                continue
            geneta = abs(getattr(data,'geneta'))
            region_check = False
            for ieta1,ieta2 in zip(self.etas_l, self.etas_r):
                if (geneta < ieta1 or geneta > ieta2): 
                    continue
                if region_check:
                    raise ValueError('This event was already assigned to an eta region.')
                region_check = True
                recen = getattr(data,'en_sr{}_ROI'.format(ireg))
                avgnoise = ( getattr(data,'noise_sr3_ROI')
                             *self.sr_area[ireg-1]/self.sr_area[2] )
                npidx = np.where(self.etas_l==ieta1)[0][0]
                x[npidx].append( [genen, geneta, recen, avgnoise] )
        return [np.array(x[i]) for i in range(len(x))]


    def nopu_calibration(self, nq=6, plot=True):
        """
        Performs the L0 (uniform eta response) and L1 (absolute scale) calibrations.
        When dealing with pile-up, 'pu_calibration()' has to be called afterwords.
        """
        import numpy as np
        from array import array as Carray
        from ROOT import TFile, TH2F
        from UserCode.HGCalMaskVisualProd.SystemUtils import EtaStr as ES

        fIn = TFile.Open(self.flags.noPUFile)
        data = fIn.Get('data')

        for ireg in [1,2,3]:
            x = self._energies_for_calibration(data, ireg)
            for ieta1,ieta2 in zip(self.etas_l, self.etas_r):
                idx = np.where(self.etas_l==ieta1)[0][0]
                idstr = 'sr{}_from{}to{}'.format(ireg, ES(ieta1), ES(ieta2))
                xq = np.percentile(x[idx], [i*100./nq for i in range(0,nq+1)], axis=0)
                #relative calibration versus eta
                hn = idstr+'_'+self.samples+'_mask'+self.mask
                htmp = TH2F('resVSeta_'+hn, 
                            ';|#eta|;#DeltaE/E', nq, Carray('d',xq[:,1]), 100,-1,1)
                for i in range(0,len(x[idx])):
                    genEn, genEta, recEn,_ = x[idx][i]
                    deltaE = (recEn/genEn) - 1.
                    htmp.Fill(genEta, deltaE)
                self.calib['L0'][idstr] = self._calibrate_spectrum(htmp,'SR%d'%ireg,
                                                                   self.label+' (PU=0)',
                                                                   func='pol2',
                                                                   plot=plot)
                htmp.Delete()

                #relative calibration versus energy
                htmp = TH2F('resVSen_'+hn,
                            ';Reconstructed energy [GeV];#DeltaE/E', nq, 
                            Carray('d',xq[:,0]), 100,-1,1)
                for i in range(0,len(x[idx])):
                    genen, geneta, recen, _ = x[idx][i]
                    recen /= (self.calib['L0'][idstr].Eval(geneta)+1.0)
                    deltaE = (recen/genen) - 1.
                    htmp.Fill(recen, deltaE)
                self.calib['L1'][idstr] = self._calibrate_spectrum(htmp,'SR%d'%ireg,
                                                                   self.label+' (PU=0)',
                                                                   func='pol1',
                                                                   plot=plot)
                htmp.Delete()
        fIn.Close()

    #correct arguments of np.percentile
    def pu_calibration(self, nq=10, plot=True):
        """
        To be used in pile-up situations only.
        It parametrizes the absolute shift in energy as function of the 
        average noise in the SR cone.
        """
        import numpy as np
        from array import array as Carray
        from ROOT import TFile, TH2F
        from UserCode.HGCalMaskVisualProd.SystemUtils import EtaStr as ES

        fIn=TFile.Open(self.flags.PUFile)
        data=fIn.Get('data')

        for ireg in [1,2,3]:
            x = self._energies_for_calibration(data, ireg)
            for ieta1,ieta2 in zip(self.etas_l, self.etas_r):
                idx = np.where(self.etas_l==ieta1)[0][0]
                idstr = 'sr{}_from{}to{}'.format(ireg, ES(ieta1), ES(ieta2))
                xq=np.percentile(x, [i*100./nq for i in range(0,nq+1)], axis=0)
                print(xq)
                #relative calibration versus eta
                resolVsNoiseFrac = TH2F('resVSnoisefrac_'+self.samples+'_'+self.mask+'_',
                                        ';<Noise> [GeV];#DeltaE [GeV]', nq, 
                                        Carray('d',xq[:,3]),50,-100,100)
                for i in range(0,len(x)):
                    genen, geneta, recen, noiseest = x[i]
                    recen /= (self.calib['L0'][idstr].Eval(geneta)+1.0)
                    recen /= (self.calib['L1'][idstr].Eval(recen)+1.0)
                    deltaE = recen - genen 
                    resolVsNoiseFrac.Fill(noiseEst, deltaE)
    
                self.calib['L2'][ireg] = self._calibrate_spectrum(resolVsNoiseFrac,
                                                                  'SR%d'%ireg,
                                                                  self.label+' (PU=140)', 
                                                                  func='pol2',
                                                                  plot=plot)
        fIn.Close()

    def save(self, name):
        if name[-4:] != '.pck':
            raise ValueError('The calibration has to be saved as a pickle file.')
        import pickle
        with open(name,'w') as cachefile:
            pickle.dump(self.calib, cachefile, pickle.HIGHEST_PROTOCOL)

File: HGCalMaskResolutionAna/Ctests/src/calibration.cc
#include "../interface/calibration.h"

Calibration::Calibration(const float mingenen_, std::vector<float> etareg_, 
			 const int nreg_, const std::string label_, 
			 const std::string samples_, const unsigned int mask_, 
			 const std::string noPUFile_, const std::string outpath_): 
  mingenen(mingenen_), etareg(etareg_), nreg(nreg_), label(label_), 
  samples(samples_), mask(mask_), noPUFile(noPUFile_), outpath(outpath_) {
  etareg_shift = VecOps(etareg).shift();
  for(int i=0; i<nreg; ++i) {
    mapstr<TF1*> tmpmap;
    calib.push_back(tmpmap);
  }
}

void Calibration::pu_calibration(const int& nq, const bool& plot) 
{
}

void Calibration::nopu_calibration(const int& nq, const bool& plot) 
{
  for(int ireg=1; ireg<=nreg; ++ireg) {      
    std::string hstr = "SR" + std::to_string(ireg);
    vec3d<float> x = Calibration::energies_for_calibration("data", ireg);
    typename std::vector<float>::const_iterator it;

    //loop over eta calibration regions
    for(it=etareg.cbegin(); it!=(etareg.cend()-1); ++it) {
      int idx = std::distance(etareg.cbegin(), it);

      float etatmp1 = *it;      
      float etatmp2 = *(it+1);
      std::string str1 = std::to_string(etatmp1).replace(1,1,"p").erase(5,10);
      std::string str2 = std::to_string(etatmp2).replace(1,1,"p").erase(5,10);
      std::string idstr = "sr" + std::to_string(ireg) + "from" + str1 + "to" + str2;

      unsigned int ss = x[idx].size();
      double xq0[ss], xq1[ss];
      double probs[nq+1];
      for(int j=0; j<nq+1; ++j)
	probs[j] = static_cast<double>(j)/nq;
 
      for(int j=0; j<ss; ++j) {
	xq0[j] = static_cast<double>(x[idx][j][0]);
	xq1[j] = static_cast<double>(x[idx][j][1]);
      }
      double quantiles0[nq+1], quantiles1[nq+1]; 
      TMath::Quantiles(ss, nq+1, xq0, quantiles0, probs, kFALSE);
      TMath::Quantiles(ss, nq+1, xq1, quantiles1, probs, kFALSE);

      //relative calibration versus eta
      std::string hn = idstr + "_" + this->samples + "_mask" + std::to_string(this->mask);
      TH2D* htmp1 = new TH2D(("resVSeta_"+hn).c_str(), ";|#eta|;#DeltaE/E", 
			     nq, quantiles1, 100, -1, 1);
      float genen, geneta, recen, deltaE;
      for(int i=0; i<ss; ++i) {
	genen = x[idx][i][0];
	geneta = x[idx][i][1];
	recen = x[idx][i][2];
	deltaE = (recen/genen) - 1.;
	htmp1->Fill(geneta, deltaE);
      }
      this->calib[0][idstr] = Calibration::calibrate_spectrum(htmp1, hstr, label+"(PU=0)",
							      "pol2", plot);
      delete htmp1;

      //relative calibration versus energy
      TH2D *htmp2 = new TH2D(("resVSen_"+hn).c_str(), ";Reco energy [GeV];#DeltaE/E", 
			     nq, quantiles0, 100, -1, 1);
      for(int i=0; i<ss; ++i) {
	genen = x[idx][i][0];
	geneta = x[idx][i][1];
	recen = x[idx][i][2];
	recen /= (this->calib[0][idstr]->Eval(geneta)+1.0);
	deltaE = (recen/genen) - 1.;
	htmp2->Fill(recen, deltaE);
      }
      this->calib[1][idstr] = Calibration::calibrate_spectrum(htmp2, hstr, label+" (PU=0)",
							      "pol1", plot);
      delete htmp2;
    }
  }
}

TF1* Calibration::calibrate_spectrum(TH2D* h, const std::string& title, 
				     const std::string& proc, const std::string& func, 
				     const bool& plot)
{
  TGraphAsymmErrors* prof = build_median_profile(h);
  prof->Fit(func.c_str());
  std::string clone_str = h->GetName() + title + "_calib";
  TF1* calibGr = static_cast<TF1*>(prof->GetListOfFunctions()->At(0)->Clone(clone_str.c_str()) );
 
  if(plot) {
    TCanvas c("c", "c", 500, 500);
    c.SetTopMargin(0.05);
    c.SetBottomMargin(0.1);
    c.SetLeftMargin(0.12);
    c.SetRightMargin(0.1);
    h->Draw("colz");
    prof->Draw("e2p");

    TLatex tex;
    tex.SetTextFont(42);
    tex.SetTextSize(0.04);
    tex.SetNDC();
    tex.DrawLatex(0.12, 0.96, "#bf{CMS} #it{simulation preliminary}");
    tex.DrawLatex(0.15, 0.88, title.c_str());
    tex.SetTextAlign(31);
    tex.DrawLatex(0.97, 0.96, proc.c_str());
    c.SaveAs( (this->outpath + h->GetName() + "_" + title + ".png").c_str() );
  }

  prof->Delete();
  return calibGr;
}

//correct SR detail
vec3d<float> Calibration::energies_for_calibration(const std::string& tname, 
					    const int& ireg) 
{
  vec3d<float> x;
  //fill energies vector with 2d vectors, one per etaregion considered in the calibration
  for(int i=0; i<this->etareg.size()-1; ++i) {
    vec2d<float> etareg_row;
    x.push_back(etareg_row);
  }

  std::vector<float> newetareg(this->etareg.cbegin(), this->etareg.cend());
  auto fill_layer_energies = [&x, &newetareg](const float& gen, const float& geta, 
					     const float& en, const float& noi) {
    bool region_check = false;
    typename std::vector<float>::const_iterator it;
    for(it = newetareg.cbegin(); it!=(newetareg.cend()-1); ++it) {
      if(geta < *it || geta > *(it+1))
	continue; 
      assert(!region_check);
      region_check = true;
      float avgnoise = noi * 1;//self.sr_area[ireg-1]/self.sr_area[2]

      std::vector<float> row = {gen, geta, en, noi}; 
      int idx = std::distance(newetareg.cbegin(), it);
      x.at(idx).push_back(row);
    }
  };
  auto cut = [](float var, float cut) {return var > cut;};

  std::string mingenen_str = "static_cast<float>(" + std::to_string(mingenen) + ")";
  ROOT::EnableImplicitMT(10);
  ROOT::RDataFrame d(tname, noPUFile);
  d.Define("abs_geneta", "abs(geneta)")
    .Define("mingenen", mingenen_str)
    .Filter(cut, {"genen", "mingenen"})
    .Foreach(fill_layer_energies, {"genen","abs_geneta","en_sr1_ROI","noise_sr3_ROI"});
  return x;
}
